ros2-foxy: OS-0 sensor hangs while activating

## Describe the bug

Driver can connect to the sensor and successfully pulls the metadata, but it hangs while the sensor is activating. The topics are visible using RQT, but data never actually gets published.

## Setup

### `docker-compose.yml` - container runtime configuration

```yaml
ouster:
    build:
      context: https://github.com/ouster-lidar/ouster-ros.git#ros2-foxy
      args:
        - ROS_DISTRO=foxy
    restart: unless-stopped
    network_mode: host
    tty: true
    stdin_open: true
    environment:
      - ROS_DOMAIN_ID=101
      - ROS_DISTRO=foxy
      - WORKSPACE=/var/lib/build # needed for /ros_entrypoint.sh to find install/setup.bash
    volumes:
      - ../scripts/ros_entrypoint.sh:/ros_entrypoint.sh
      - ../config/sensors/ouster_params.yml:/var/lib/build/src/ouster-ros/ouster-ros/config/parameters.yaml
    entrypoint:
      - /ros_entrypoint.sh
    command:
      - ros2
      - launch
      - ouster_ros
      - sensor.independent.launch.py
      - viz:=false
```

### `ouster_params.yml` - sensor configuration

```yaml
ouster:
  os_sensor:
    ros__parameters:
      sensor_hostname: 10.0.0.32
      udp_dest: 10.0.0.60
      lidar_mode: 2048x10
      timestamp_mode: TIME_FROM_ROS_TIME
      udp_profile_lidar: LEGACY
      lidar_port: 7501
      imu_port: 7502
      metadata: ''

  os_cloud:
    ros__parameters:
      tf_prefix: left_os_zero
      timestamp_mode: TIME_FROM_ROS_TIME
```

### `ros_entrypoint.sh` - Docker container entrypoint

```bash
#!/bin/bash
set -e

# setup ros environment
ROS_SETUP="/opt/ros/$ROS_DISTRO/setup.bash"
if [ -f "$ROS_SETUP" ]; then
    source "$ROS_SETUP"
    echo "Sourced $(readlink -f $ROS_SETUP)"
fi

# setup ROS packages
PKG_SETUP="$WORKSPACE/install/setup.bash"
if [ -f "$PKG_SETUP" ]; then
    source "$PKG_SETUP"
    echo "Sourced $(readlink -f $PKG_SETUP)"
fi

exec "$@"
```

## Results
### Command
```
docker-compose -f ./docker/docker-compose.yml --build up ouster
```
### Output
```shell
Sourced /opt/ros/foxy/setup.bash
Sourced /var/lib/build/install/setup.bash
[INFO] [launch]: All log files can be found below /var/lib/build/.ros/log/2023-03-10-20-16-24-748088-rapid-dev-box-1-1
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [os_sensor-1]: process started with pid [54]
[INFO] [os_cloud-2]: process started with pid [56]
[INFO] [os_image-3]: process started with pid [58]
[os_image-3] [INFO] [1678479384.872404521] [ouster.os_image]: contacting get_metadata service; attempt no: 1/10
[os_cloud-2] [INFO] [1678479384.877239482] [ouster.os_cloud]: contacting get_metadata service; attempt no: 1/10
[os_sensor-1] [INFO] [1678479385.097626349] [ouster.os_sensor]: TIME_FROM_ROS_TIME timestamp mode specified. IMU and pointcloud messages will use ros time
[os_sensor-1] [INFO] [1678479385.097917648] [ouster.os_sensor]: Will send UDP data to 10.0.0.60
[os_sensor-1] [INFO] [1678479385.190119356] [ouster.os_sensor]: Sensor 10.0.0.32 configured successfully
[os_sensor-1] [INFO] [1678479385.190318752] [ouster.os_sensor]: Starting sensor 10.0.0.32 initialization...
[os_sensor-1] [2023-03-10 20:16:25.190] [ouster::sensor] [info] initializing sensor: 10.0.0.32 with ports: 7501/7502
[os_sensor-1] [INFO] [1678479386.649603744] [ouster.os_sensor]: ouster client version: 0.7.1+unknown-release
[os_sensor-1] product: OS-0-64-U02, sn: 122226001747, firmware rev: v2.3.1
[os_sensor-1] lidar mode: 2048x10, lidar udp profile: LEGACY
[os_sensor-1] [INFO] [1678479386.649715174] [ouster.os_sensor]: No metadata file was specified, using: 10.0.0-metadata.json
[os_sensor-1] [INFO] [1678479386.650035131] [ouster.os_sensor]: Wrote sensor metadata to 10.0.0-metadata.json
[os_sensor-1] [INFO] [1678479386.651548360] [ouster.os_sensor]: reset service created
[os_sensor-1] [INFO] [1678479386.653285624] [ouster.os_sensor]: get_metadata service created
[os_image-3] [INFO] [1678479386.653573389] [ouster.os_image]: waiting for get_metadata service to respond; attempt no: 1/10
[os_cloud-2] [INFO] [1678479386.653738287] [ouster.os_cloud]: waiting for get_metadata service to respond; attempt no: 1/10
[os_sensor-1] [INFO] [1678479386.654336994] [ouster.os_sensor]: get_config service created
[os_sensor-1] [INFO] [1678479386.655261165] [ouster.os_sensor]: set_config service created
[INFO] [launch.user]: os_sensor activating...
[os_image-3] [INFO] [1678479386.657114815] [ouster.os_image]: retrieved sensor metadata!
[os_cloud-2] [INFO] [1678479386.657131660] [ouster.os_cloud]: retrieved sensor metadata!
```

## Platform (please complete the following information):**
- Ouster Sensor? OS-0
- Ouster Firmware Version?
```shell
"sensor_info": 
    {
        "base_pn": "",
        "base_sn": "",
        "build_date": "2022-06-08T01:26:30Z",
        "build_rev": "v2.3.1",
        "image_rev": "ousteros-image-prod-aries-v2.3.1+20220608012528.patch-v2.3.x",
        "initialization_id": 8247304,
        "prod_line": "OS-0-64-U02",
        "prod_pn": "840-103574-06",
        "prod_sn": "122226001747",
        "proto_rev": "",
        "status": "RUNNING"
    }
```
- ROS version/distro? ROS2-foxy
- Operating System? Linux
- Machine Architecture? x86 and arm64

I think I have the same issue. The driver works well in WSL with Ubuntu Jammy and Humble (as long as the UDP ports are forwarded), but does not in a Ubuntu Jammy Docker container. The sensor gets configured (i.e. switches from operating mode "Standby" to "Normal") but then hangs at:
`[component_container_mt-1] [INFO] [1679653979.326010969] [ouster.os_image]: waiting for get_metadata service to respond; attempt no: 4/10`


<details>
  <summary>See the full log</summary>
  <br>
  $ros2 launch ouster_ros sensor.launch.xml sensor_hostname:=192.168.1.60 lidar_port:=7502 imu_port:=7503  
  <br>    
    
  [INFO] [launch]: All log files can be found below /home/grisi/.ros/log/2023-03-24-11-11-56-349559-molisens-349
  [INFO] [launch]: Default logging verbosity is set to INFO
  [INFO] [component_container_mt-1]: process started with pid [361]
  [INFO] [bash-2]: process started with pid [363]
  [INFO] [bash-3]: process started with pid [365]
  [INFO] [bash-4]: process started with pid [367]
  [INFO] [static_transform_publisher-5]: process started with pid [369]
  [INFO] [static_transform_publisher-6]: process started with pid [373]
  [static_transform_publisher-5] [INFO] [1679652716.688072914] [ouster.stp_sensor_imu]: Spinning until stopped - publishing transform
  [static_transform_publisher-5] translation: ('0.000000', '0.000000', '0.000000')
  [static_transform_publisher-5] rotation: ('0.000000', '0.000000', '0.000000', '1.000000')
  [static_transform_publisher-5] from 'os_sensor' to 'os_imu'
  [static_transform_publisher-6] [INFO] [1679652716.693022295] [ouster.stp_sensor_lidar]: Spinning until stopped - publishing transform
  [static_transform_publisher-6] translation: ('0.000000', '0.000000', '0.000000')
  [static_transform_publisher-6] rotation: ('0.000000', '0.000000', '0.000000', '1.000000')
  [static_transform_publisher-6] from 'os_sensor' to 'os_lidar'
  [component_container_mt-1] [INFO] [1679652716.874694400] [ouster.os_container]: Load Library: /home/grisi/MOLISENS/molisens_ws/install/ouster_ros/lib/libos_sensor_component.so
  [component_container_mt-1] [INFO] [1679652716.885163776] [ouster.os_container]: Found class: rclcpp_components::NodeFactoryTemplate<ouster_ros::OusterSensor>
  [component_container_mt-1] [INFO] [1679652716.885294091] [ouster.os_container]: Instantiate class: rclcpp_components::NodeFactoryTemplate<ouster_ros::OusterSensor>
  [INFO] [launch_ros.actions.load_composable_nodes]: Loaded node '/ouster/os_sensor' in container '/ouster/os_container'
  [component_container_mt-1] [INFO] [1679652716.898561801] [ouster.os_container]: Load Library: /home/grisi/MOLISENS/molisens_ws/install/ouster_ros/lib/libos_cloud_component.so
  [component_container_mt-1] [INFO] [1679652716.901720241] [ouster.os_container]: Found class: rclcpp_components::NodeFactoryTemplate<ouster_ros::OusterCloud>
  [component_container_mt-1] [INFO] [1679652716.901792870] [ouster.os_container]: Instantiate class: rclcpp_components::NodeFactoryTemplate<ouster_ros::OusterCloud>
  [component_container_mt-1] [INFO] [1679652716.908755480] [ouster.os_cloud]: contacting get_metadata service; attempt no: 1/10
  [component_container_mt-1] [INFO] [1679652718.364315653] [ouster.os_sensor]: Will use automatic UDP destination
  [component_container_mt-1] [INFO] [1679652718.448792975] [ouster.os_sensor]: Sensor 192.168.1.60 configured successfully
  [component_container_mt-1] [INFO] [1679652718.448907176] [ouster.os_sensor]: Starting sensor 192.168.1.60 initialization...
  [component_container_mt-1] [2023-03-24 11:11:58.448] [ouster::sensor] [info] initializing sensor: 192.168.1.60 with ports: 7502/7503
  [component_container_mt-1] [INFO] [1679652719.716175158] [ouster.os_sensor]: ouster client version: 0.7.1+202c536-release
  [component_container_mt-1] product: OS-1-64-U02, sn: 992201000519, firmware rev: v2.3.0
  [component_container_mt-1] lidar mode: 1024x10, lidar udp profile: RNG19_RFL8_SIG16_NIR16_DUAL
  [component_container_mt-1] [INFO] [1679652719.716262591] [ouster.os_sensor]: No metadata file was specified, using: 192.168.1-metadata.json
  [component_container_mt-1] [INFO] [1679652719.716401862] [ouster.os_sensor]: Wrote sensor metadata to 192.168.1-metadata.json
  [component_container_mt-1] [INFO] [1679652719.717029177] [ouster.os_sensor]: reset service created
  [component_container_mt-1] [INFO] [1679652719.717335979] [ouster.os_sensor]: get_metadata service created
  [component_container_mt-1] [INFO] [1679652719.717430748] [ouster.os_cloud]: waiting for get_metadata service to respond; attempt no: 1/10
  [component_container_mt-1] [INFO] [1679652719.717735073] [ouster.os_sensor]: get_config service created
  [component_container_mt-1] [INFO] [1679652719.717975130] [ouster.os_sensor]: set_config service created
  [bash-2] Transitioning successful
  [component_container_mt-1] [INFO] [1679652719.719185517] [ouster.os_cloud]: retrieved sensor metadata!
  [bash-3] Transitioning successful
  [INFO] [launch_ros.actions.load_composable_nodes]: Loaded node '/ouster/os_cloud' in container '/ouster/os_container'
  [component_container_mt-1] [INFO] [1679652719.760760674] [ouster.os_container]: Load Library: /home/grisi/MOLISENS/molisens_ws/install/ouster_ros/lib/libos_image_component.so
  [component_container_mt-1] [INFO] [1679652719.762364334] [ouster.os_container]: Found class: rclcpp_components::NodeFactoryTemplate<ouster_ros::OusterImage>
  [component_container_mt-1] [INFO] [1679652719.762402606] [ouster.os_container]: Instantiate class: rclcpp_components::NodeFactoryTemplate<ouster_ros::OusterImage>
  [component_container_mt-1] [INFO] [1679652719.765768648] [ouster.os_image]: contacting get_metadata service; attempt no: 1/10
  [component_container_mt-1] [INFO] [1679652719.765842619] [ouster.os_image]: waiting for get_metadata service to respond; attempt no: 1/10
  [INFO] [bash-3]: process has finished cleanly [pid 365]
  [INFO] [bash-2]: process has finished cleanly [pid 363]
  [bash-4] QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-grisi'
  [bash-4] [INFO] [1679652721.856418288] [rviz2]: Stereo is NOT SUPPORTED
  [bash-4] [INFO] [1679652721.856576954] [rviz2]: OpenGl version: 4.5 (GLSL 4.5)
  [bash-4] [INFO] [1679652721.921919205] [rviz2]: Stereo is NOT SUPPORTED
  [component_container_mt-1] [WARN] [1679652721.988104380] [ouster.os_cloud]: New subscription discovered on topic '/ouster/points', requesting incompatible QoS. No messages will be sent to it. Last incompatible policy: RELIABILITY_QOS_POLICY
  [bash-4] [INFO] [1679652722.897403148] [rviz2]: Stereo is NOT SUPPORTED
  [component_container_mt-1] [INFO] [1679652729.766439155] [ouster.os_image]: waiting for get_metadata service to respond; attempt no: 2/10
  [component_container_mt-1] [INFO] [1679652739.767111198] [ouster.os_image]: waiting for get_metadata service to respond; attempt no: 3/10
  [component_container_mt-1] [INFO] [1679652749.767764013] [ouster.os_image]: waiting for get_metadata service to respond; attempt no: 4/10
  [component_container_mt-1] [INFO] [1679652759.768630278] [ouster.os_image]: waiting for get_metadata service to respond; attempt no: 5/10
  [component_container_mt-1] [INFO] [1679652769.769938338] [ouster.os_image]: waiting for get_metadata service to respond; attempt no: 6/10
  [component_container_mt-1] [INFO] [1679652779.770803561] [ouster.os_image]: waiting for get_metadata service to respond; attempt no: 7/10
  [component_container_mt-1] [INFO] [1679652789.771498287] [ouster.os_image]: waiting for get_metadata service to respond; attempt no: 8/10
  [component_container_mt-1] [INFO] [1679652799.772305887] [ouster.os_image]: waiting for get_metadata service to respond; attempt no: 9/10
  [component_container_mt-1] [INFO] [1679652809.772905806] [ouster.os_image]: waiting for get_metadata service to respond; attempt no: 10/10
  [component_container_mt-1] [ERROR] [1679652819.773809215] [ouster.os_image]: get_metadata service timed out or interrupted
  [ERROR] [launch_ros.actions.load_composable_nodes]: Failed to load node 'os_image' of type 'ouster_ros::OusterImage' in container '/ouster/os_container': Component constructor threw an exception: get_metadata service timed out or interrupted
  [component_container_mt-1] [ERROR] [1679652819.777422482] [ouster.os_container]: Component constructor threw an exception: get_metadata service timed out or interrupted

## Previous commit

To try the `sensor.composite.launch.py` launch file, I had to change the Docker build context to use the same commit as when I first created the issue.

```docker
ouster:
    build:
      context: https://github.com/ouster-lidar/ouster-ros.git#4fbf576eebcfeeb00d7e9a42f016e84152ca09c5
      args:
        - ROS_DISTRO=foxy
    ...
    command:
      - ros2
      - launch
      - ouster_ros
      - sensor.composite.launch.py # changed to other launch file
      - viz:=false
```

New error messages:

```shell
[component_container_mt-1] [ERROR] [1680715918.147608892] [ouster.os_container]: Component constructor threw an exception: tf_prefix
[ERROR] [launch_ros.actions.load_composable_nodes]: Failed to load node 'os_cloud' of type 'ouster_ros::OusterCloud' in container '/ouster/os_container': Component constructor threw an exception: tf_prefix
...
[component_container_mt-1] [ERROR] [1680715919.221592574] [ouster.os_sensor]: exception thrown while configuring the sensor, details: sensor_hostname
```

## Current `ros2-foxy` branch

New error message, using `sensor.independent.launch.py`

```shell
[ERROR] [launch_ros.actions.lifecycle_node]: Failed to make transition 'TRANSITION_CONFIGURE' for LifecycleNode '/ouster/os_sensor'
[os_sensor-1] [ERROR] [1680725907.046408387] [ouster.os_sensor]: exception thrown while configuring the sensor, details: mtp_dest
[os_sensor-1] [WARN] [1680725907.046743305] []: Error occurred while doing error handling.
```

--- a/ouster-ros/CMakeLists.txt
+++ b/ouster-ros/CMakeLists.txt
@@ -24,6 +24,10 @@ find_package(tf2_eigen REQUIRED)
 
 # ==== Options ====
 add_compile_options(-Wall -Wextra)
+if(NOT DEFINED CMAKE_CXX_STANDARD)
+  set(CMAKE_CXX_STANDARD 17)
+  set(CMAKE_CXX_STANDARD_REQUIRED ON)
+endif()
 option(CMAKE_POSITION_INDEPENDENT_CODE "Build position independent code." ON)
 
 

